### PR TITLE
Add location `SpanTuple`s

### PR DESCRIPTION
**Pull request by CohenArthur**

### CODE REVIEW

File: src/context.rs [hunk 1]
```diff
@@ -89,6 +92,7 @@ impl Context {
     pub fn new() -> Context {
         let mut ctx = Context {
             debug_mode: false,
+            code: String::new(), // FIXME: Is this the correct behavior?
```
Comment (Skallwar) [inline src/context.rs:95]: I think an `Option` would be a good choice here

File: src/error/mod.rs [hunk 2]
```diff
@@ -74,61 +77,45 @@ impl ErrKind {
...
+                    Location::new(max(loc.start().line() - 3, 1), loc.start().column());
+                let before_ctx = SpanTuple::new(before_start, loc.start().clone());
+                before_ctx.emit('|', input);
+            }
+
+            loc.emit("x".yellow(), input);
+
+            let after_end = Location::new(loc.end().line() + 3, loc.end().column());
+            let after_ctx = SpanTuple::new(loc.end().clone(), after_end);
+            after_ctx.emit('|', input);
```
Comment (Skallwar) [inline src/error/mod.rs:105]: Lots of magic values here

File: src/error/mod.rs [hunk 3]
```diff
@@ -192,7 +179,7 @@ impl From<nom::Err<Error>> for Error {
 impl<'i> nom::error::ParseError<LocatedSpan<&'i str>> for Error {
     fn from_error_kind(span: LocatedSpan<&'i str>, _: nom::error::ErrorKind) -> Error {
         // FIXME: Add location here
```
Comment (Skallwar) [inline src/error/mod.rs:181]: Is this still relevant?

Comment (CohenArthur) [reply src/error/mod.rs:181]: The locations for parsing errors aren't satisfactory right now, so I think we should keep this still

File: src/error/mod.rs (hunk 2 above)
Comment (CohenArthur) [reply src/error/mod.rs]: Fixed in 350b010

File: src/context.rs (hunk 1 above)
Comment (CohenArthur) [reply src/context.rs]: Done in 05e98bd

File: src/error/mod.rs (hunk 3 above)
Comment (CohenArthur) [reply src/error/mod.rs]: I tweaked the error message so that it reflects the idea better

**Comment by CohenArthur** [comment]
I think this PR is becoming a bit big. Two main issues remain:
- [ ] When printing the context, some "fake lines" are duplicated:
```rust
error_type: Typechecker

    5 | type NewType[T](inner: T);
    6 | // This function has more context than the previous one. For example, it also
    7 | // declares a new type!
    8 |
    8 x func another_type_error(b: int) -> string {
    9 x
   10 x 	b
   11 x }
   11 |
   12 |
   13 | // Waaaay more context here!
   14 | we_add_a_new_variable = 15

test.jk:8:1: invalid type returned in function `another_type_error`: expected type string, found type int
```
Here, on lines 8 and 11 which appear twice and empty
- [ ] When printing whole lines in the context, we are missing the last character of the last line:
```rust
error_type: Typechecker

    1 x func tc_error(a: int) -> float {
    2 x 	a
    3 x }
    3 |
    4 |
    5 | type NewType[T](inner: T);
    6 | // This function has more context than the previous one. For example, it als

test.jk:1:1: invalid type returned in function `tc_error`: expected type float, found type int
```
Here, as pointed out earlier, line 6 is supposed to say "For example, it also"

**Review by Skallwar — APPROVED** [review]

**Comment by CohenArthur** [comment]
> I think this PR is becoming a bit big. Two main issues remain:
> 
>     * [x]  When printing the context, some "fake lines" are duplicated:
> 
> 
> ```rust
> error_type: Typechecker
> 
>     5 | type NewType[T](inner: T);
>     6 | // This function has more context than the previous one. For example, it also
>     7 | // declares a new type!
>     8 |
>     8 x func another_type_error(b: int) -> string {
>     9 x
>    10 x 	b
>    11 x }
>    11 |
>    12 |
>    13 | // Waaaay more context here!
>    14 | we_add_a_new_variable = 15
> 
> test.jk:8:1: invalid type returned in function `another_type_error`: expected type string, found type int
> ```
> 
> Here, on lines 8 and 11 which appear twice and empty
> 
>     * [x]  When printing whole lines in the context, we are missing the last character of the last line:
> 
> 
> ```rust
> error_type: Typechecker
> 
>     1 x func tc_error(a: int) -> float {
>     2 x 	a
>     3 x }
>     3 |
>     4 |
>     5 | type NewType[T](inner: T);
>     6 | // This function has more context than the previous one. For example, it als
> 
> test.jk:1:1: invalid type returned in function `tc_error`: expected type float, found type int
> ```
> 
> Here, as pointed out earlier, line 6 is supposed to say "For example, it also"

Both of those issues were fixed in #498

**Comment by CohenArthur** [comment]
I think we can merge this and keep working on adding locations for various other instructions

**Pull request merged**